Concurrent Modification during first start up

````
20:58:50.448 [pool-1-thread-1] ERROR java.lang.Thread -- ═╦═⚀ Fatal implementation error in or by using Thread[#29,pool-1-thread-1,5,main]: UncaughtException found!
20:58:50.448 [pool-1-thread-1] ERROR java.lang.Thread --  ╚════⚀ ConcurrentModificationException
20:58:50.448 [pool-1-thread-1] ERROR java.lang.Thread -- =====================================
20:58:50.448 [pool-1-thread-1] ERROR java.lang.Thread -- Fatal implementation error in or by using Thread[#29,pool-1-thread-1,5,main]: UncaughtException found!
org.openbase.jul.exception.FatalImplementationErrorException: Fatal implementation error in or by using Thread[#29,pool-1-thread-1,5,main]: UncaughtException found!
	at org.openbase.jul.schedule.AbstractExecutorService$1.uncaughtException(AbstractExecutorService.java:88)
	at java.base/java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:695)
	at java.base/java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:690)
	at java.base/java.lang.Thread.dispatchUncaughtException(Thread.java:2902)
Caused by: java.util.ConcurrentModificationException: null
	at java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1095)
	at java.base/java.util.ArrayList$Itr.next(ArrayList.java:1049)
	at org.openbase.bco.device.hass.communication.websocket.HassWebsocketConnection._set_connectionState_$lambda$5$lambda$4(HassWebsocketConnection.kt:272)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
20:58:50.449 [pool-1-thread-1] ERROR java.lang.Thread -- =====================================
20:58:50.449 [pool-1-thread-88] INFO org.openbase.bco.device.hass.communication.websocket.HassWebsocketConnection -- subscribe on: WSSubscription(commandType=subscribe_events, eventType=FLOOR_UPDATE, eventProcessor=org.openbase.bco.device.hass.manager.service.location.LocationSynchronizer$$Lambda/0x00000088014f2d68@5ae227d)
20:58:50.450 [pool-1-thread-1] ERROR org.openbase.jul.schedule.AbstractExecutorService$1 -- ═╦═⚀ Fatal implementation error in or by using Thread[#29,pool-1-thread-1,5,main]: UncaughtException found!
20:58:50.450 [pool-1-thread-1] ERROR org.openbase.jul.schedule.AbstractExecutorService$1 --  ╚════⚀ ConcurrentModificationException
20:58:50.450 [pool-1-thread-1] ERROR org.openbase.jul.schedule.AbstractExecutorService$1 -- =====================================
20:58:50.450 [pool-1-thread-1] ERROR org.openbase.jul.schedule.AbstractExecutorService$1 -- Fatal implementation error in or by using Thread[#29,pool-1-thread-1,5,main]: UncaughtException found!
org.openbase.jul.exception.FatalImplementationErrorException: Fatal implementation error in or by using Thread[#29,pool-1-thread-1,5,main]: UncaughtException found!
	at org.openbase.jul.schedule.AbstractExecutorService$1.uncaughtException(AbstractExecutorService.java:88)
	at java.base/java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:695)
	at java.base/java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:690)
	at java.base/java.lang.Thread.dispatchUncaughtException(Thread.java:2902)
Caused by: java.util.ConcurrentModificationException: null
	at java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1095)
	at java.base/java.util.ArrayList$Itr.next(ArrayList.java:1049)
	at org.openbase.bco.device.hass.communication.websocket.HassWebsocketConnection._set_connectionState_$lambda$5$lambda$4(HassWebsocketConnection.kt:272)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

--- a/src/main/kotlin/org/openbase/bco/device/hass/communication/websocket/HassWebsocketConnection.kt
+++ b/src/main/kotlin/org/openbase/bco/device/hass/communication/websocket/HassWebsocketConnection.kt
@@ -55,6 +55,7 @@ class HassWebsocketConnection(
     private val requestMapLock: ReentrantReadWriteLock = ReentrantReadWriteLock()
     private val connectionStateLock: ReentrantReadWriteLock = ReentrantReadWriteLock()
     private val connectionStateCondition: Condition = connectionStateLock.writeLock().newCondition()
+    private val subscriptionsLock = ReentrantReadWriteLock()
     private val subscriptions: MutableList<WSSubscription> = emptyList<WSSubscription>().toMutableList()
 
 
@@ -72,10 +73,12 @@ class HassWebsocketConnection(
 
                 if (value == ConnectionState.State.CONNECTED) {
                     GlobalCachedExecutorService.execute {
-                        subscriptions.forEach { subscription ->
-                            sendSubscriptionRequest(subscription)
-                                .await()
-                                .also { LOGGER.info("subscribe on: $subscription") }
+                        subscriptionsLock.read {
+                            subscriptions.forEach { subscription ->
+                                sendSubscriptionRequest(subscription)
+                                    .await()
+                                    .also { LOGGER.info("subscribe on: $subscription") }
+                            }
                         }
                     }
                 }
@@ -98,18 +101,20 @@ class HassWebsocketConnection(
     fun subscribe(
         subscription: WSSubscription,
     ) {
-        subscriptions
-            .find { it == subscription}
-            ?.let { LOGGER.error("Already subscribed to ${subscription.commandType}") }
-            ?: subscriptions.add(subscription).also {
-                connectionStateLock.read {
-                    if (connectionState == ConnectionState.State.CONNECTED) {
-                        sendSubscriptionRequest(subscription)
-                            .await()
-                            .also { LOGGER.info("subscribe on: $subscription") }
+        subscriptionsLock.write {
+            subscriptions
+                .find { it == subscription}
+                ?.let { LOGGER.error("Already subscribed to ${subscription.commandType}") }
+                ?: subscriptions.add(subscription).also {
+                    connectionStateLock.read {
+                        if (connectionState == ConnectionState.State.CONNECTED) {
+                            sendSubscriptionRequest(subscription)
+                                .await()
+                                .also { LOGGER.info("subscribe on: $subscription") }
+                        }
                     }
                 }
-            }
+        }
     }
 
     override fun activate() {
@@ -206,7 +211,11 @@ class HassWebsocketConnection(
                 }
                 "event" -> {
                     JsonUtils.gson.fromJson(jsonResult, SubscriptionEvent::class.java).also { result ->
-                        subscriptions.find { it.eventType?.eventTypeName == result.event.eventType }?.eventProcessor?.invoke(result.event)
+                        subscriptionsLock.read {
+                            subscriptions.find { it.eventType?.eventTypeName == result.event.eventType }?.eventProcessor?.invoke(
+                                result.event
+                            )
+                        }
                     }
                     return
                 }

